oil_lang/expr_eval.py

```diff
@@ -579,64 +633,243 @@ def _EvalUnary(self, node):
+            elif op == Id.Arith_Star:
+              return value.Float(left.f * right.i)
+            elif op == Id.Arith_Slash:
+              if right.i == 0:
+                raise ZeroDivisionError()
+
+              return value.Float(left.f / right.i)
+            else:
+              raise NotImplementedError(op)
+
+          elif rcase(value_e.Float):
+            right = cast(value__Float, UP_right)
+            if op == Id.Arith_Plus:
+              return value.Float(left.f + right.f)
+            elif op == Id.Arith_Minus:
+              return value.Float(left.f - right.f)
+            elif op == Id.Arith_Star:
+              return value.Float(left.f * right.f)
+            elif op == Id.Arith_Slash:
+              if right.f == 0.0:
+                raise ZeroDivisionError()
+
+              return value.Float(left.f / right.f)
+            else:
+              raise NotImplementedError(op)
 
-    if node.op.id == Id.Arith_Plus:
-      return self._ToNumber(left) + self._ToNumber(right)
-    if node.op.id == Id.Arith_Minus:
-      return self._ToNumber(left) - self._ToNumber(right)
-    if node.op.id == Id.Arith_Star:
-      return self._ToNumber(left) * self._ToNumber(right)
+          else:
+            raise error.InvalidType('Expected Int or Float', loc.Missing())
+
+      else:
+          raise error.InvalidType('Expected Int or Float', loc.Missing())
+
+  def _ArithDivideInt(self, left, right):
+    # type: (value_t, value_t) -> value__Int
+    left_i = self._ValueToInteger(left)
+    right_i = self._ValueToInteger(right)
+    if right_i == 0:
+      raise ZeroDivisionError()
+
+    return value.Int(left_i // right_i)
+
+  def _ArithModulus(self, left, right):
+    # type: (value_t, value_t) -> value__Int
+    left_i = self._ValueToInteger(left)
+    right_i = self._ValueToInteger(right)
+    if right_i == 0:
+      raise ZeroDivisionError()
+
+    return value.Int(left_i % right_i)
+
+  def _ArithExponentiate(self, left, right):
+    # type: (value_t, value_t) -> value__Int
+    left_i = self._ValueToInteger(left)
+    right_i = self._ValueToInteger(right)
+    return value.Int(left_i ** right_i)
+
+  def _ArithBitwise(self, left, right, op):
+    # type: (value_t, value_t, Id_t) -> value__Int
+    left_i = self._ValueToInteger(left)
+    right_i = self._ValueToInteger(right)
+
+    if op == Id.Arith_Amp:
+      return value.Int(left_i & right_i)
+    elif op == Id.Arith_Pipe:
+      return value.Int(left_i | right_i)
+    elif op == Id.Arith_Caret:
+      return value.Int(left_i ^ right_i)
+    elif op == Id.Arith_DGreat:
+      return value.Int(left_i >> right_i)
+    elif op == Id.Arith_DLess:
+      return value.Int(left_i << right_i)
+
+    raise NotImplementedError()
+
+  def _ArithLogical(self, left, right, op):
+    # type: (value_t, value_t, Id_t) -> value__Bool
+    UP_left = left
+    UP_right = right
+
+    with tagswitch(left) as lcase:
+      if lcase(value_e.Bool):
```

Comment:
I wonder if we can somehow switch left and right at the same time, I'll think about that ...

The tags are small integers so technically we could shift them into a single integer like (16 bit left tag, 16 bit right tag)

Reply:
ah yes that would back following the logic here much easier